enhancements/ingress/dynamic-config-manager.md

```diff
@@ -0,0 +1,469 @@
+a considerably large amount of memory.  See
+https://gist.github.com/frobware/2b527ce3f040797909eff482a4776e0b for an
+analysis of the potential memory impact for different choices of balancing
+algorithm, number of threads, number of backends, and number of pre-allocated
+server slots per backend.
+
+To avoid operational overhead, avoid adding new APIs, and minimize overhead of
+enabling DCM, we will initially omit blueprint routes and only pre-allocate 1
+server slot per backend.  This makes DCM an implementation detail; ideally it is
+completely invisible to the cluster-admin, except that the router will fork
+fewer processes and have more accurate metrics and balancing.  Later, we can
+enhance DCM by using newer capabilities in HAProxy 2.y to add arbitrarily many
+servers dynamically, add backends dynamically, and update certificates
+dynamically.
+
+### Risks and Mitigations
+
+Dynamic Config Manager has some known defects:
+
+- [OCPBUGS-7466 No prometheus haproxy metrics present for route created by dynamicConfigManager](https://issues.redhat.com/browse/OCPBUGS-7466)
+- [OCPBUGSM-20868 Sticky cookies take long to start working if config manager is enabled](https://issues.redhat.com/browse/OCPBUGSM-20868)
+- [NE-1815 Fix implementation gaps discovered during the smoke tests](https://issues.redhat.com/browse/NE-1815)
+
+We intend to fix all of these defects before enabling DCM by default as GA.
+However, DCM could have additional, unknown defects.  We will mitigate this risk
+through E2E tests, payload tests, and working with partners to test DCM.
+
+### Drawbacks
+
+#### Memory overhead
+
+DCM also has some up-front memory overhead.  For this reason, DCM will be
+initially enabled with a minimum configuration of 1 pre-allocated server slot
+per backend and without blueprint routes.  Eventually, DCM will be enhanced to
+use newer HAProxy capabilities to perform more configuration without
+pre-allocating servers or backends, thus preventing reloads in more cases
+without adding any additional up-front memory overhead.
+
+#### Redundancy with respect to Istio/Envoy
+
+DCM is similar in concept to [Envoy
+xDS](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/dynamic_configuration).
+Both enable a control plane (openshift-router or Istio) to configure a data
+plane (HAProxy or Envoy, respectively) without restarting processes.  Investing
+in DCM could be considered a redundant effort when Istio/Envoy already exists
+and avoids the problem.  However, many customers depend on OpenShift router for
+its performance and reliability and will continue using it indefinitely.  For
+these customers, DCM could improve OpenShift router's performance and
+reliability without forcing a change in APIs (from Route API to Gateway API) or
+proxies (from HAProxy to Envoy).
+
+## Design Details
+
+### Open Questions [optional]
+
+#### Can we use HAProxy's control socket to add backends and certificates dynamically?
+
+Answer: **?**.
```

Comment:
Backends - no. Certificates - seems like [yes](https://www.haproxy.com/documentation/haproxy-runtime-api/reference/new-ssl-cert/).

Reply:
Yeah, it does seem that dynamic backends are are long discussed but never implemented possibility.  Sadface.